main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1387,6 +1390,91 @@ object Defaults extends BuildCommon {
     }
     analysisResult.analysis
   }
+  lazy val previousAnalysis = Def.inputKey[String]("gets previous analysis")
+  lazy val previousAnalysisParser = Def.setting {
+    Parsers.StringBasic
+  }
+  lazy val previousAnalysisCommand = Command.single("previousAnalysis") { (state, definition) =>
+    val responseId = state.currentCommand.flatMap(_.execId)
+    state.log.warn(s"definition content: $definition, and id: $responseId")
+    import sbt.internal.langserver.TextDocumentPositionParams
+    import sbt.internal.langserver.codec.JsonProtocol._
+    import sjsonnew.support.scalajson.unsafe.{ Parser => JsonParser, Converter }
+    val input =
+      Converter.fromJsonUnsafe[TextDocumentPositionParams](JsonParser.parseUnsafe(definition))
+    state.log.err(input.toString)
+    val updatedState = Project.updateCurrent(state).reload
+    val aggrAnal = {
+      val e = Project.extract(updatedState)
+      val skey = (Keys.compile in Compile).scopedKey
+      val tasks = e.structure.data.scopes
+        .map { scope =>
+          e.structure.data.get(scope, skey.key)
+        }
+        .collect {
+          case Some(t) => Aggregation.KeyValue(skey, t)
+        }
+        .toSeq
+      state.log.warn(s"tasks $tasks")
+      val complete = Aggregation.timedRun(updatedState, tasks, sbt.std.Transform.DummyTaskMap(Nil))
+      state.log.warn("here")
+      val r = complete.results.toEither.toOption.map { results =>
+        state.log.warn("and there")
+        results.map(_.value)
+      }
+      r
+    }.getOrElse(Seq.empty)
+    state.log.warn(s"aggregated analysis $aggrAnal")
+
+    val analysis = {
+      val compile = Keys.compile in Compile
+      Project.runTask(compile, updatedState, true).collect {
+        case (newstate @ _, Value(analysis)) =>
+          analysis
+      }
+    }.orElse {
+        state.log.warn("analysis not found")
+        None
+      }
+      .collect { case a: Analysis => a }
+    state.log.warn(s"analysis: $analysis")
+    analysis
+      .foreach { analysis =>
+        analysis.relations.allSources.headOption.foreach { whatever =>
+          state.log.warn(s"found $whatever")
+        }
+      }
+    val commandAnalysis = state.get(Keys.analysis)
+    state.log.warn(s"command analysis: $commandAnalysis")
+    val previousCompile = Project.runTask(Keys.previousCompile in Compile, updatedState).collect {
+      case (newstate @ _, Value(v)) if v.analysis.isPresent => v.analysis.get
+    }
+    state.log.warn(s"previous analysis: $previousCompile")
+    val cmdSrc = state.currentCommand.flatMap(_.source)
+    import sbt.internal.langserver.{ Location, Position, Range }
+    val params = Location(input.textDocument.uri, Range(Position(0, 0), Position(0, 10)))
+    val cname = cmdSrc.map(_.channelName).get
+    val chnl = StandardMain.exchange.channels.collectFirst {
+      case c if c.name == cname => c
+    }
+    chnl.foreach(_.publishEvent(params, responseId))
```

Comment:
If possible, we should use event logging here.

Reply:
What is event logging?

Reply:
Calling `infoEvent` on ManagedLogger, which then gets relayed to CommandExchange.

Reply:
sorry, I don't get it.